src/mds/SessionMap.h

```diff
@@ -215,20 +216,54 @@ class Session : public RefCountedObject {
     ceph_assert(!info.prealloc_inos.empty());
 
     if (ino) {
-      if (info.prealloc_inos.contains(ino))
+      if (info.prealloc_inos.contains(ino)) {
 	info.prealloc_inos.erase(ino);
-      else
+	if (delegated_inos.contains(ino))
+	  delegated_inos.erase(ino);
+      } else {
 	ino = 0;
+      }
     }
     if (!ino) {
-      ino = info.prealloc_inos.range_start();
-      info.prealloc_inos.erase(ino);
+      /* Grab first prealloc_ino that isn't delegated */
+      for (interval_set<inodeno_t>::const_iterator p = info.prealloc_inos.begin(); p != info.prealloc_inos.end(); ++p) {
+	for (auto i = (*p).first ; i < (*p).first+(*p).second ; i += 1) {
+	  if (!delegated_inos.contains(i)) {
```

Comment:
interval_set::contains(T i, T *pstart=0, T *plen=0) const.

we can pass start, len to contain(...), make the iteration more efficient``

Reply:
Good idea. I'll mix that in.